src/main/java/org/sagebionetworks/bridge/models/schedules2/adherence/study/StudyAdherenceReportGenerator.java

```diff
@@ -0,0 +1,272 @@
+                        studyStream.addEntry(numDays, oneDay);
+                        
+                        String eventId = oneDay.getStartEventId();
+                        eventTimestamps.put(eventId, state.getEventTimestampById(eventId));
+                    } else {
+                        // Note for the report the events and sessions that are not applicable to this user
+                        unsetEventIds.add(oneDay.getStartEventId());
+                        unscheduledSessions.add(oneDay.getSessionName());
+                    }
+                }
+            }
+        }
+        // study-wide progress
+        ParticipantStudyProgress progression = calculateProgress(state, ImmutableList.of(studyStream));
+        
+        // Break this study stream down into weeks. TreeMap sorts the weeks by week number 
+        Map<Integer, StudyReportWeek> weekMap = new TreeMap<>();
+        for (Map.Entry<Integer, List<EventStreamDay>> entry : studyStream.getByDayEntries().entrySet()) {
+            int week = entry.getKey() / 7;
+            // Day of week can be negative if the week was negative... just flip it
+            int dayOfWeek = Math.abs(entry.getKey() % 7);
+            
+            StudyReportWeek oneWeek = weekMap.get(week);
+            if (oneWeek == null) {
+                oneWeek = new StudyReportWeek();
+                oneWeek.setStartDate(studyStartDate.plusDays(week*7));
+                oneWeek.setWeek(week+1);
+                
+                weekMap.put(week, oneWeek);    
+            }
+            List<EventStreamDay> days = oneWeek.getByDayEntries().get(dayOfWeek);
+            days.addAll(entry.getValue());    
+        };
+
+        Collection<StudyReportWeek> weeks = weekMap.values();
+        StudyReportWeek currentWeek = null;
+        
+        for (StudyReportWeek oneWeek : weeks) {
+            calculateRowsAndLabels(oneWeek, todayLocal);
+            
+            // If there’s a day in this week that is “today”, set a flag for all day entries on that day
+            LocalDate firstDayOfWeek = oneWeek.getStartDate();
+            LocalDate lastDayOfWeek = oneWeek.getStartDate().plusDays(7);
+            if (BridgeUtils.isLocalDateInRange(firstDayOfWeek, lastDayOfWeek, todayLocal)) {
+                currentWeek = oneWeek;
+                for (List<EventStreamDay> days : oneWeek.getByDayEntries().values()) {
+                    for (EventStreamDay oneDay : days) {
+                        if (oneDay.getStartDate().isEqual(todayLocal)) {
+                            days.forEach(day -> day.setToday(true));
+                        }
+                    }
+                }
+            }
+            if (BridgeUtils.isLocalDateInRange(oneWeek.getStartDate(), null, todayLocal)) {
+                int weekAdh = AdherenceUtils.calculateAdherencePercentage(oneWeek.getByDayEntries());
+                oneWeek.setAdherencePercent(weekAdh);
+            }
```

Comment:
Why does this work?

Reply:
Because the range of startDate- matches every week up to the current week, but after that, todayLocal is BEFORE the range, it doesn't match and, so adherence is not calculated for the weeks after the current week. Added comment and test to verify this. It's the correct behavior.